include/eosio/vm/host_function.hpp

```diff
@@ -214,24 +214,32 @@ namespace eosio { namespace vm {
    template <typename T, typename WAlloc>
    constexpr auto resolve_result(T&& res, WAlloc*) -> std::enable_if_t<
          !(std::is_pointer_v<T> || std::is_reference_v<T>)&&std::is_same_v<to_wasm_t<T>, i32_const_t>, i32_const_t> {
-      return i32_const_t{ *(uint32_t*)&res };
+      if constexpr (std::is_same_v<std::decay_t<T>, bool>) {
+         return i32_const_t{ res };
+      } else {
+         static_assert(sizeof(T) == sizeof(uint32_t));
+         return i32_const_t{ *(uint32_t*)&res };
+      }
    }
 
    template <typename T, typename WAlloc>
    constexpr auto resolve_result(T&& res, WAlloc*) -> std::enable_if_t<
          !(std::is_pointer_v<T> || std::is_reference_v<T>)&&std::is_same_v<to_wasm_t<T>, i64_const_t>, i64_const_t> {
+      static_assert(sizeof(T) == sizeof(uint64_t));
       return i64_const_t{ *(uint64_t*)&res };
    }
 
    template <typename T, typename WAlloc>
    constexpr auto resolve_result(T&& res, WAlloc*) -> std::enable_if_t<
          !(std::is_pointer_v<T> || std::is_reference_v<T>)&&std::is_same_v<to_wasm_t<T>, f32_const_t>, f32_const_t> {
+      static_assert(sizeof(T) == sizeof(uint32_t));
       return f32_const_t{ *(uint32_t*)&res };
```

Comment:
Fixing the casts will remove the need for the static_assert.  In addition, only signed X/unsigned X are safe to convert like this.  It's UB for other integral types even if they happen to have the same size.  See [basic.lval]

Reply:
The real fix would need a memcpy (or in C++ 20, bit_cast) since these do bitcasts. e.g. double to uint64_t.

Reply:
`to_softfloat32(res)`